File: docs/scripts/learner.js
const TOKEN_PATTERN = /[A-Za-z']+|\d+|[.,!?]/g;
const GIBBERISH_SYLLABLES = [
  'za',
  'gru',
  'plo',
  'fen',
  'zib',
  'vor',
  'quar',
  'nib',
  'mal',
  'tre',
];

export class MarkovLearner {
  constructor(rng = Math.random) {
    this.rng = rng;
  }

  nextReply(messages) {
    const tokens = messages.flatMap((message) => tokenize(message.text));
    if (tokens.length < 16) {
      return gibberish(this.rng);
    }
    const chain = buildChain(tokens);
    return composeSentence(chain, tokens, this.rng);
  }
}

export function tokenize(text) {
  if (!text) {
    return [];
  }
  return (text.toLowerCase().match(TOKEN_PATTERN) ?? []);
}

function buildChain(tokens) {
  const transitions = new Map();
  for (let i = 0; i < tokens.length - 1; i += 1) {
    const current = tokens[i];
    const next = tokens[i + 1];
    if (!transitions.has(current)) {
      transitions.set(current, []);
    }
    transitions.get(current).push(next);
  }
  return transitions;
}

function composeSentence(chain, corpusTokens, rng) {
  if (!corpusTokens.length) {
    return gibberish(rng);
  }
  const start = corpusTokens[Math.floor(rng() * corpusTokens.length)];
  const length = Math.floor(rng() * 15) + 6; // 6 to 20 words
  const words = [start];
  for (let i = 1; i < length; i += 1) {
    const options = chain.get(words[words.length - 1]);
    if (!options || options.length === 0) {
      break;
    }
    const choice = options[Math.floor(rng() * options.length)];
    words.push(choice);
  }
  let sentence = formatSentence(words);
  if (!sentence.endsWith('.') && !sentence.endsWith('!') && !sentence.endsWith('?')) {
    sentence += '.';
  }
  return sentence;
}

function gibberish(rng) {
  const wordCount = Math.floor(rng() * 5) + 3; // 3 to 7 words
  const words = [];
  for (let i = 0; i < wordCount; i += 1) {
    const syllables = Math.floor(rng() * 3) + 2; // 2 to 4 syllables
    let word = '';
    for (let s = 0; s < syllables; s += 1) {
      const choice = GIBBERISH_SYLLABLES[Math.floor(rng() * GIBBERISH_SYLLABLES.length)];
      word += choice;
    }
    words.push(word);
  }
  let sentence = words.join(' ');
  sentence = sentence.charAt(0).toUpperCase() + sentence.slice(1);
  const endings = ['?', '!', '...'];
  sentence += endings[Math.floor(rng() * endings.length)];
  return sentence;
}

function formatSentence(tokens) {
  if (!tokens.length) {
    return '';
  }
  let sentence = tokens[0];
  for (let i = 1; i < tokens.length; i += 1) {
    const token = tokens[i];
    if (['.', ',', '!', '?'].includes(token)) {
      sentence += token;
    } else {
      sentence += ` ${token}`;
    }
  }
  sentence = sentence.charAt(0).toUpperCase() + sentence.slice(1);
  return sentence;
}

File: src/learner.py
"""Simple conversational learner that stores dialogues in JSON."""

from __future__ import annotations

import json
import random
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, Iterable, List, Sequence

# Regular expression to split text into word-like tokens while keeping punctuation.
_TOKEN_PATTERN = re.compile(r"[A-Za-z']+|\d+|[.,!?]")


@dataclass
class Message:
    """Represents a single utterance in the conversation log."""

    speaker: str
    text: str
    timestamp: str

    def as_dict(self) -> Dict[str, str]:
        return {"speaker": self.speaker, "text": self.text, "timestamp": self.timestamp}


class ConversationStore:
    """Handles loading and saving conversation history to a JSON file."""

    def __init__(self, path: Path) -> None:
        self.path = path
        self.path.parent.mkdir(parents=True, exist_ok=True)
        if self.path.exists():
            try:
                data = json.loads(self.path.read_text(encoding="utf-8"))
            except json.JSONDecodeError as exc:  # pragma: no cover - defensive branch
                raise ValueError(f"Conversation file {self.path} is not valid JSON") from exc
        else:
            data = {"messages": []}
            self._write(data)
        self._messages: List[Message] = [
            Message(**message) for message in data.get("messages", []) if _validate_message(message)
        ]

    @property
    def messages(self) -> Sequence[Message]:
        return tuple(self._messages)

    def append(self, speaker: str, text: str) -> Message:
        message = Message(speaker=speaker, text=text, timestamp=_current_timestamp())
        self._messages.append(message)
        self._write({"messages": [msg.as_dict() for msg in self._messages]})
        return message

    def _write(self, payload: Dict[str, Iterable[Dict[str, str]]]) -> None:
        self.path.write_text(json.dumps(payload, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")


class MarkovChatbot:
    """A tiny chatbot that learns a Markov model from prior conversation."""

    def __init__(self, store: ConversationStore, seed: int | None = None) -> None:
        self.store = store
        self.random = random.Random(seed)

    def respond(self, user_text: str) -> str:
        user_text = user_text.strip()
        if not user_text:
            raise ValueError("Cannot learn from empty user input.")
        self.store.append("user", user_text)
        reply = self._generate_reply()
        self.store.append("bot", reply)
        return reply

    def _generate_reply(self) -> str:
        messages = self.store.messages
        tokens = [token for message in messages for token in _tokenize(message.text)]
        if len(tokens) < 16:
            return _gibberish(self.random)
        chain = _build_chain(tokens)
        return _compose_sentence(chain, tokens, self.random)


def _validate_message(message: Dict[str, str]) -> bool:
    return {
        "speaker",
        "text",
        "timestamp",
    }.issubset(message)


def _current_timestamp() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def _tokenize(text: str) -> List[str]:
    return [match.group(0).lower() for match in _TOKEN_PATTERN.finditer(text)]


def _build_chain(tokens: Sequence[str]) -> Dict[str, List[str]]:
    transitions: Dict[str, List[str]] = {}
    for current, nxt in zip(tokens, tokens[1:]):
        transitions.setdefault(current, []).append(nxt)
    return transitions


def _compose_sentence(chain: Dict[str, List[str]], corpus_tokens: Sequence[str], rng: random.Random) -> str:
    if not corpus_tokens:
        return _gibberish(rng)
    start = rng.choice(corpus_tokens)
    length = rng.randint(6, 20)
    words = [start]
    for _ in range(length - 1):
        options = chain.get(words[-1])
        if not options:
            break
        words.append(rng.choice(options))
    sentence = _format_sentence(words)
    if sentence[-1] not in ".!?":
        sentence += "."
    return sentence


_GIBBERISH_SYLLABLES = [
    "za",
    "gru",
    "plo",
    "fen",
    "zib",
    "vor",
    "quar",
    "nib",
    "mal",
    "tre",
]


def _gibberish(rng: random.Random) -> str:
    word_count = rng.randint(3, 7)
    words = []
    for _ in range(word_count):
        syllables = rng.randint(2, 4)
        word = "".join(rng.choice(_GIBBERISH_SYLLABLES) for _ in range(syllables))
        words.append(word)
    sentence = " ".join(words)
    sentence = sentence.capitalize()
    sentence += rng.choice(["?", "!", "..."])
    return sentence


def _format_sentence(tokens: Sequence[str]) -> str:
    if not tokens:
        return ""
    sentence = tokens[0]
    for token in tokens[1:]:
        if token in {".", ",", "!", "?"}:
            sentence += token
        else:
            sentence += " " + token
    return sentence.capitalize()


__all__ = ["ConversationStore", "MarkovChatbot", "Message"]
